crates/verify/src/bytecode/zksync.rs

```diff
@@ -0,0 +1,485 @@
+        &project.settings.settings,
+    )
+    .await;
+
+    if shell::is_json() {
+        sh_println!("{}", serde_json::to_string(&json_results)?)?;
+    }
+
+    Ok(())
+}
+
+fn match_bytecodes(
+    local_bytecode: &[u8],
+    onchain_bytecode: &[u8],
+    local_bytecode_hash: BytecodeHash,
+    onchain_bytecode_hash: BytecodeHash,
+) -> Option<VerificationType> {
+    // 1. Try full match
+    if local_bytecode == onchain_bytecode {
+        // If the bytecode_hash = 'none' in Config. Then it's always a partial match according to
+        // sourcify definitions. Ref: https://docs.sourcify.dev/docs/full-vs-partial-match/.
+        if local_bytecode_hash == BytecodeHash::None {
+            return Some(VerificationType::Partial);
+        }
+
+        Some(VerificationType::Full)
+    } else {
+        extract_and_compare_bytecode(
+            local_bytecode,
+            onchain_bytecode,
+            local_bytecode_hash,
+            onchain_bytecode_hash,
+        )
+        .then_some(VerificationType::Partial)
+    }
+}
+
+fn extract_and_compare_bytecode(
+    local_bytecode: &[u8],
+    onchain_bytecode: &[u8],
+    local_hash_type: BytecodeHash,
+    onchain_hash_type: BytecodeHash,
+) -> bool {
+    // Only compare valid bytecodes
+    if local_bytecode.len() % 64 != 32 || onchain_bytecode.len() % 64 != 32 {
+        return false;
+    }
+
+    let local_bytecode_without_hash = extract_metadata_hash(local_bytecode, local_hash_type);
+    let onchain_bytecode_without_hash = extract_metadata_hash(onchain_bytecode, onchain_hash_type);
+
+    // In order to be a match, bytecodes need to either be equal or just differ on an extra word
+    // of 0s added to achieve odd 32 byte word count
+    let min_len =
+        std::cmp::min(local_bytecode_without_hash.len(), onchain_bytecode_without_hash.len());
+
+    let (local_bytecode_without_remainder, local_remainder) =
+        local_bytecode_without_hash.split_at(min_len);
+    let (onchain_bytecode_without_remainder, onchain_remainder) =
+        onchain_bytecode_without_hash.split_at(min_len);
+
+    let no_remainders = local_remainder.is_empty() && onchain_remainder.is_empty();
+    let local_has_padding = local_remainder.len() == 32 && local_remainder.iter().all(|b| *b == 0);
+    let onchain_has_padding =
+        onchain_remainder.len() == 32 && onchain_remainder.iter().all(|b| *b == 0);
+    if !(no_remainders || local_has_padding || onchain_has_padding) {
+        return false;
+    }
+
+    local_bytecode_without_remainder == onchain_bytecode_without_remainder
+}
+
+fn extract_metadata_hash(bytecode: &[u8], hash_type: BytecodeHash) -> &[u8] {
+    let hash_len = match hash_type {
+        BytecodeHash::None => 0,
+        BytecodeHash::Keccak256 => 32,
+        BytecodeHash::Ipfs => {
+            // Get the last two bytes of the bytecode to find the length of CBOR metadata
+            let metadata_len = &bytecode[bytecode.len() - 2..];
+            let metadata_len: usize =
```

Comment:
If you're adding `2` which I presume is the length header size, this is changing the variable `metadata_len` defined above which actually contains the actual metadata length without the header/trailer. Maybe define a different variable to better emphasize this difference. `metadata_with_header_len` or something.